### ISSUE TITLE
problem

### ISSUE THREAD

**Issue by Ulricalang** [YMJS-Irfan/DP-FedSAM#2]
Hello, I have a question. When I was running the residual 34, there was a nan value in the global test accuracy. I don't know where there was a problem.

**Comment by YMJS-Irfan** [comment]
Thanks for focusing on our work! This issue hasn't happened in our experimental settings; the detailed description can see in our CVPR paper. 
And for some scenarios where larger C can lead to the gradient explosion phenomenon, there was a nan value in the global test accuracy. This issue may be caused by a large variance magnitude of DP random noise due to large C.
<img width="543" alt="image" src="https://github.com/YMJS-Irfan/DP-FedSAM/assets/67813720/f890ed0d-4d5f-4c15-ad16-1f9e0321a130">

Some explanation and experimental setup can be found in **Section** 6.1 of our CVPR paper :
<img width="714" alt="image" src="https://github.com/YMJS-Irfan/DP-FedSAM/assets/67813720/757bb964-03dd-4e85-8b86-a41881b17f3b">

**Comment by YMJS-Irfan** [comment]
> Hello, I have a question. When I was running the residual 34, there was a nan value in the global test accuracy. I don't know where there was a problem.

Thanks for focusing on our work! This issue hasn't happened in our experimental settings; the detailed description can see in our CVPR paper. 
And for some scenarios where larger C can lead to the gradient explosion phenomenon, there was a nan value in the global test accuracy. This issue may be caused by a large variance magnitude of DP random noise due to large C.
<img width="543" alt="image" src="https://github.com/YMJS-Irfan/DP-FedSAM/assets/67813720/f890ed0d-4d5f-4c15-ad16-1f9e0321a130">

Some explanation and experimental setup can be found in **Section** 6.1 of our CVPR paper :
<img width="714" alt="image" src="https://github.com/YMJS-Irfan/DP-FedSAM/assets/67813720/757bb964-03dd-4e85-8b86-a41881b17f3b">

**Comment by Ulricalang** [comment]
Is the global_test_loss value too large because of this problem?

**Comment by YMJS-Irfan** [comment]
> Is the global_test_loss value too large because of this problem?

Yes, maybe this issue causes. We also found this phenomenon under some hyper-parameter settings.

**Comment by Ulricalang** [comment]
I also want to ask, I tried to add some layers to the model, which was normal during training, but the loss value was unusually large during testing. Is it possible that this is also related to the noise value? Did you try it during the test?

**Comment by Ulricalang** [comment]
Thank you very much for your answer. I wish you all the best.

**Comment by YMJS-Irfan** [comment]
> I also want to ask, I tried to add some layers to the model, which was normal during training, but the loss value was unusually large during testing. Is it possible that this is also related to the noise value? Did you try it during the test?

Furthermore, we find that this issue may possibly occur in the EMNIST data. Maybe you can try it on CIFAR-10/100.
Best wish.

**Comment by ssissw** [comment]
> > Hello, I have a question. When I was running the residual 34, there was a nan value in the global test accuracy. I don't know where there was a problem.
> 
> Thanks for focusing on our work! This issue hasn't happened in our experimental settings; the detailed description can see in our CVPR paper. And for some scenarios where larger C can lead to the gradient explosion phenomenon, there was a nan value in the global test accuracy. This issue may be caused by a large variance magnitude of DP random noise due to large C. <img alt="image" width="543" src="https://private-user-images.githubusercontent.com/67813720/256145918-f890ed0d-4d5f-4c15-ad16-1f9e0321a130.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU2NjU2NTcsIm5iZiI6MTcwNTY2NTM1NywicGF0aCI6Ii82NzgxMzcyMC8yNTYxNDU5MTgtZjg5MGVkMGQtNGQ1Zi00YzE1LWFkMTYtMWY5ZTAzMjFhMTMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE5VDExNTU1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUyYWM5YjA4MDhjODQ1MDc1NGJhNGJkNzUyMGNiODQ3NWFiMDBkNzZjNWVmMTkzNjM3Y2U0OTU5OTNkMDM5Y2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.khS_SgnJUY91VXkKTIBpSZ0BRPMfN02IgxZbbLE_Oww">
> 
> Some explanation and experimental setup can be found in **Section** 6.1 of our CVPR paper : <img alt="image" width="714" src="https://private-user-images.githubusercontent.com/67813720/256146447-757bb964-03dd-4e85-8b86-a41881b17f3b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU2NjU2NTcsIm5iZiI6MTcwNTY2NTM1NywicGF0aCI6Ii82NzgxMzcyMC8yNTYxNDY0NDctNzU3YmI5NjQtMDNkZC00ZTg1LThiODYtYTQxODgxYjE3ZjNiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE5VDExNTU1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE4NDNmNTRhMzU4Y2MzODc3YWZkMzAyOWJiNTQzZTc2Zjg2YmRiMDE2OGMxMGY4YmQ0YWMyZjE0NDFjMmQ1OWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.w3ytEti7Z0IFXg_w2v3vIUv4eStanNmVKO8w_kGSPxs">

I wonder why the var of noise is (σ^2 * C^2 / m), since the sensitivity of each client update is bounded by (C / m), may the var should be (σ^2 * C^2 / m^2)?
